### PR TITLE
Added more tests for fifo_fast

### PR DESCRIPTION
**Pull request by forumulator**
Three things:

1. I've made slight mods to fifo_fast, adding some limits, please review them.

2. The underflow test doesn't work, because for some reason it takes 2 yield clock.posedge statements to update values at certain times, thus it gives an assertion error instead of expected ValueError. I've also commented in the code. Need to debug that later.

3. As we discussed, we need some means of error checking. I too like the idea of an ExceptionSignal common to rhea, which we could then include in various interfaces or use separately.

Your thoughts?

### CODE REVIEW

**Comment by cfelton** [comment]
One of the things we can do is use `pytest.mark.xfail`, meaning that we know the test is failing and we will get to it in the future.

**Comment by cfelton** [comment]
I comment before looking at the code, you used the `xfail`, perfect.

File: test/test_cores/test_fifo/test_ffifo.py [hunk 1]
```diff
@@ -80,16 +167,213 @@ def tbstim():
...
+            for ii in range(5):
+                yield clock.posedge
+            raise StopSimulation
+        return tbdut, tbclk, tbstim
+    for kk in range(100):
+        with pytest.raises(ValueError):
+            run_testbench(_bench_fifo_overflow)
+
+# test underflow
+# should give a ValueError
```
Comment (cfelton) [inline test/test_cores/test_fifo/test_ffifo.py]: These pre-function comments can move to the doc-string, one place for global comments/docs for a function.

**Comment by cfelton** [comment]
In general this works for now, specific tests can still be setup to verify an exception is raised:

```python
with pytest.raises(ValueError):
    run_testbenchd(bench, args=args)
```

This will allow the test to produce the error conditions and validate the exception is being raised.  This is very helpful in development, because developers are informed of these error early on.

Right now, the errors mainly occur on the `nvacant` and `ntenant`, I need to think about it more but the range of these counters can be extended so the errors don't occur here but can occur else where.  The whole idea of ExceptionSignals will take a while to flush out.

Looks good, I am going to merge this, an issue should be created for the "todo" items so we don't forget about them.

File: rhea/cores/fifo/fifo_fast.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def fifo_fast(clock, reset, fbus, use_srl_prim=False):
         # the FIFO, only a read address is accounted.
         @always(clock.posedge)
         def rtl_srl_in():
-            if srlce:
+            if srlce and (not fbus.full):
```
Comment (cfelton) [inline rhea/cores/fifo/fifo_fast.py]: This change might be backed out later, the SRL fifo was organized so the SRL FIFO would be inferred, I don't think these extra checks can be used in the SRL fifo, the normal synchronous FIFO can have the additional checks.

File: rhea/cores/fifo/fifo_fast.py [hunk 3]
```diff
@@ -97,7 +97,10 @@ def rtl_fifo():
 
         elif fbus.write and not fbus.read:
             fbus.empty.next = False
-            if not fbus.empty:
+            # because addr starts from 0,
+            # there should not be an increment
+            # the first time 
+            if (not fbus.empty) and (not fbus.full):
```
Comment (cfelton) [inline rhea/cores/fifo/fifo_fast.py]: I am confused, I don't understand how the previous tests would have passed?  Something is not lining up, either the previous tests were invalid or this change is incorrect.

Comment (forumulator) [inline rhea/cores/fifo/fifo_fast.py]: The previous tests were perfectly valid. Since we start the empty queue at addr 0 and not -1, we would need to stay one step behind in the increments ,so the 
```python
if (not fbus.empty) condition is needed
```
I added the other 
```python
if (not fbus.full)
```
because in other instances the fifo continues to increment the `addr` counter even after the fifo is full. This is sometimes prevented because the `nvacant` decrement below zero throws up an exception, but if there were some code, that were to ignore the exception, or keep writing after the fifo is full, `addr` would otherwise end on a totally out of bounds end value, this prevents that. The previous tests you wrote didn't check for out of bounds, and even if they did, they would catch the exception, so this issue doesn't usually occur (It felt semantically more correct too, tell me if you disagree).

If you want me to remove the changes, of course, I will.

File: rhea/cores/fifo/fifo_fast.py (hunk 2 above)
Comment (forumulator) [inline rhea/cores/fifo/fifo_fast.py]: Sure, I'll remove this one after you tell me if you want the one below this removed too.

File: rhea/cores/fifo/fifo_fast.py (hunk 3 above)
Comment (cfelton) [inline rhea/cores/fifo/fifo_fast.py]: These need to be removed or verified that an SRL structured FIFO is created with the changes.  

If we remove the extra checks the tests can still verify that the expected exception is thrown when an overflow conditions occurs.

Did you get a response on the mailing-list?  I don't recall.  Either way, I don't think it affects this issue, an exception is still raise the exception can be verified as thrown or not thrown.  The setting of the value shouldn't matter because the execution (simulation) will stop because of the exception.

Comment (forumulator) [inline rhea/cores/fifo/fifo_fast.py]: Will remove them for now with the possibility of verifying SRL fifo and adding them later in a separate PR

**Comment by forumulator** [comment]
Reverted the changes to `fifo_fast`. I still think those 2 conditions I inserted would be useful, but that is the topic of another discussion now. This PR now only contains the added tests. If the Travis passes(it will take a long time, though :) ), I think you can now merge it into master.

**Comment by cfelton** [comment]
@forumulator thanks, yes I for the `fifo_fast` we need to do some homework and make sure the `fifo_fast` still infers the SRL type FIFO, since the SRL FIFO is limited  resources I think the additional checks will break the inference.  Two ways to check this: 1) look at the vendor synthesis guides; 2) (better) use `rhea.build` to compile the `fifo_fast` for various targets and verify before the changes and after the changes.
